flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/DefaultRescaleManager.java

```diff
@@ -0,0 +1,243 @@
+
+    private final Temporal initializationTime;
+    private final Supplier<Temporal> clock;
+
+    @VisibleForTesting final Duration scalingIntervalMin;
+    @VisibleForTesting @Nullable final Duration scalingIntervalMax;
+
+    private final RescaleManager.Context rescaleContext;
+
+    private boolean rescaleScheduled = false;
+
+    @VisibleForTesting final Duration maxTriggerDelay;
+
+    /**
+     * {@code triggerFuture} is used to allow triggering a scheduled callback. Rather than
+     * scheduling the callback itself, the callback is just chained with the future. The completion
+     * of the future is then scheduled which will, as a consequence, run the callback as part of the
+     * scheduled operation.
+     *
+     * <p>{@code triggerFuture} can be used to trigger the callback even earlier (before the
+     * scheduled delay has passed). See {@link #onTrigger()}.
+     */
+    private CompletableFuture<Void> triggerFuture;
+
+    DefaultRescaleManager(
+            Temporal initializationTime,
+            RescaleManager.Context rescaleContext,
+            Duration scalingIntervalMin,
+            @Nullable Duration scalingIntervalMax,
+            Duration maxTriggerDelay) {
+        this(
+                initializationTime,
+                Instant::now,
+                rescaleContext,
+                scalingIntervalMin,
+                scalingIntervalMax,
+                maxTriggerDelay);
+    }
+
+    @VisibleForTesting
+    DefaultRescaleManager(
+            Temporal initializationTime,
+            Supplier<Temporal> clock,
+            RescaleManager.Context rescaleContext,
+            Duration scalingIntervalMin,
+            @Nullable Duration scalingIntervalMax,
+            Duration maxTriggerDelay) {
+        this.initializationTime = initializationTime;
+        this.clock = clock;
+
+        this.maxTriggerDelay = maxTriggerDelay;
+        this.triggerFuture = FutureUtils.completedVoidFuture();
+
+        Preconditions.checkArgument(
+                scalingIntervalMax == null || scalingIntervalMin.compareTo(scalingIntervalMax) <= 0,
+                "scalingIntervalMax should at least match or be longer than scalingIntervalMin.");
+        this.scalingIntervalMin = scalingIntervalMin;
+        this.scalingIntervalMax = scalingIntervalMax;
+
+        this.rescaleContext = rescaleContext;
+    }
+
+    @Override
+    public void onChange() {
+        runInContextMainThread(
+                () -> {
+                    if (this.triggerFuture.isDone()) {
+                        this.triggerFuture =
+                                scheduleOperationWithTrigger(this::evaluateChangeEvent);
+                    }
+                });
+    }
+
+    @Override
+    public void onTrigger() {
+        runInContextMainThread(
+                () -> {
+                    if (!this.triggerFuture.isDone()) {
+                        this.triggerFuture.complete(null);
```

Comment:
As we complete the future here does it mean the `thenRun` callback is also executed in the same thread?

Reply:
Great question. There are two scenarios which we have to consider for chained future utility methods like `thenRun`:
1. The future is not completed, yet, when chaining callback to the future (through methods like `thenRun`). In this case, the callback will be executed in the thread that called `complete` on this future. In our case, that's the main thread (either through the `scheduleOperation` call or the `runInContextMainThread` call).
2. The future is completed before chaining additional callbacks to the future. In that case, the callback will be executed right away in the thread that calls the chained method (i.e. `thenRun` in our case).

The 2. scenario does not happen because we're chaining the callback in `scheduleOperationWithTrigger` before sharing the future through `triggerFuture`. And the execution of these methods happens sequentially in the main thread, anyway, so that we don't have to worry of pre-mature completion of the future.

Reply:
Perfect, thanks for the explanation.